engine/src/main/java/com/datatorrent/stram/util/JarHelper.java

```diff
@@ -0,0 +1,168 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.stram.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.JarURLConnection;
+import java.net.URL;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.security.CodeSource;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.compress.archivers.ArchiveException;
+import org.apache.commons.compress.archivers.ArchiveOutputStream;
+import org.apache.commons.compress.archivers.ArchiveStreamFactory;
+import org.apache.commons.compress.archivers.jar.JarArchiveEntry;
+import org.apache.commons.io.IOUtils;
+
+public class JarHelper
+{
+  private static final Logger logger = LoggerFactory.getLogger(JarHelper.class);
+
+  private static final Map<URL, String> sourceToJar = new HashMap<>();
+
+  public static String getJar(Class<?> jarClass)
+  {
+    String jar = null;
+    final CodeSource codeSource = jarClass.getProtectionDomain().getCodeSource();
+    if (codeSource != null) {
+      URL location = codeSource.getLocation();
+      jar = sourceToJar.get(location);
+      if (jar == null) {
+        // don't create jar file from folders multiple times
+        if ("jar".equals(location.getProtocol())) {
+          try {
+            location = ((JarURLConnection)location.openConnection()).getJarFileURL();
+          } catch (IOException e) {
+            throw new AssertionError("Cannot resolve jar file for " + jarClass, e);
+          }
+        }
+        if ("file".equals(location.getProtocol())) {
+          jar = location.getFile();
+          final File dir = new File(jar);
+          if (dir.isDirectory()) {
+            try {
+              File temp = File.createTempFile("apex-", ".jar");
+              new JarHelper(temp).createJar(dir);
```

Comment:
That should be fine or maybe have another utility class that does the Jar creation. JarHelper sounds generic and  a host for other functionality in future, not sure tying an instance down to this specific functionality would be the best for the future. You could, for example, even expose the createJar from a folder functionality from the helper.